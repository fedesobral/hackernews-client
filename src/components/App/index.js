import React, { Component } from 'react';
// import logo from './logo.svg';
import './index.css';
import Button from '../Button'
import Table from '../Table'
import Search from '../Search'

import {DEFAULT_QUERY,
        DEFAULT_PAGE,
        DEFAULT_HPP,
        PATH_BASE,
        PATH_SEARCH,
        PARAM_SEARCH,
        PARAM_PAGE,
        PARAM_HPP } from '../../constants';


class App extends Component {
  constructor(props){
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      isLoading: false
    };

    this.needsToSearchTopstories = this.needsToSearchTopstories.bind(this);
    this.setSearchTopstories = this.setSearchTopstories.bind(this);
    this.fetchSearchTopstories = this.fetchSearchTopstories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    // this.onSort = this.onSort.bind(this);
  }

  needsToSearchTopstories(searchTerm){
    return !this.state.results[searchTerm];
  }

  setSearchTopstories(result){
    const {hits, page} = result;
    const {searchKey, results } = this.state;
    const oldHits = results && results[searchKey] 
      ? results[searchKey].hits 
      : [];
    const updatedHits = [...oldHits, ...hits];
    this.setState((state, props) => {
      return {
        results: {...results, [searchKey]: {hits: updatedHits, page}},
        isLoading: false 
      };
    });
  }

  fetchSearchTopstories(searchTerm, page){
    const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`;
    this.setState((state, props) => ({ isLoading: true }));
    return fetch(url)
    .then(response => response.json())
    .then(result => this.setSearchTopstories(result));
  }

  onDismiss(id){
    const isNotId = item => item.objectID !== id;
    this.setState((state, props) => {
      const {searchKey, results} = state;
      const {hits, page} = results[searchKey];
      return { 
        results: {
          ...results, 
          [searchKey]: { hits: hits.filter(isNotId), page } 
        }
      }
    });
  }

  onSearchChange(event){
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event){
    const {searchTerm} = this.state;
    this.setState((state, props) =>  ({searchKey: searchTerm}));
    if(this.needsToSearchTopstories(searchTerm)){
      this.fetchSearchTopstories(searchTerm, DEFAULT_PAGE);
    }
    event.preventDefault();
  }

  componentDidMount() {
    const {searchTerm} = this.state
    this.setState((state, props) =>  ({searchKey: searchTerm}));
    this.fetchSearchTopstories(searchTerm, DEFAULT_PAGE);
  }

  render() {
    const { results, searchTerm, searchKey, isLoading } = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || DEFAULT_PAGE;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className='page'>
        <div className='interactions'>
          <Search
          value={searchTerm}
          onChange={this.onSearchChange}
          onSubmit={this.onSearchSubmit} >
          Search
          </Search>
        </div>
        <Table 
            list={list}
            pattern={searchTerm}
            onDismiss={this.onDismiss} />
        <div className='interactions'>
          <ButtonWithLoading 
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopstories(searchKey, page + 1)}>
            More
          </ButtonWithLoading>
        </div>
      </div>
      );
  }
}

App.defaultProps = {
  fetchInitialData: () => {
   
  }
}

const Loading = () => <div>Loading...</div>

const withLoading = (Component) => ({ isLoading, ...rest }) =>{ 
  return isLoading 
    ? <Loading />
    : <Component { ...rest } />
}
const ButtonWithLoading = withLoading(Button);

export default App;

// export {
//   Button,
//   Search,
//   Table
// }

/// React app using map and arrow function
// class App extends Component {
//   render() {
//     return (
//       <div className='App'>
//       { list.map(item =>
//         <div key={item.objectID}>
//           <span>
//             <a href={item.url}>{item.title}</a>
//           </span>
//           <span>{item.author}</span>
//           <span>{item.num_comments}</span>
//           <span>{item.points}</span>
//         </div>
//         )
//       }
//       </div>
//     );
//   }
// }
