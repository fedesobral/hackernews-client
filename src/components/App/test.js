import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount, render } from 'enzyme';
import { shallowToJson, mountToJson, renderToJson } from 'enzyme-to-json';
import App from '../App';

describe('App', () => {

	it('renders', () => {
	  const div = document.createElement('div');
	  ReactDOM.render(<App />, div);
	});

	test('snapshots', () => {
		const createNodeMock = function () {
		  // You can return anything from this function.
		  // For example:
		  return {
		    focus() {
		      // Do nothing
		    }
		  }
		}
		const component = mount(
			<App isLoading={false} />
			, {createNodeMock}
		);

		let tree = mountToJson(component); //component.toJSON();
		expect(tree).toMatchSnapshot();

		component.setState({ isLoading: true });
		component.update();
		tree = mountToJson(component); 
		expect(tree).toMatchSnapshot();
	});

	// it('pagination', () => {
	// 	const element = mount(<App />);
	// 	element.props.fetchInitialData();
	// 	element.update();
	// 	expect(element.find('.table-row').length).toBe(100);
	// });
});
