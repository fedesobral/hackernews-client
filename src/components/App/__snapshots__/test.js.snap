exports[`App snapshots 1`] = `
<App
  fetchInitialData={[Function]}
  isLoading={false}>
  <div
    className="page">
    <div
      className="interactions">
      <Search
        onChange={[Function]}
        onSubmit={[Function]}
        value="redux">
        <form
          onSubmit={[Function]}>
          <input
            onChange={[Function]}
            type="text"
            value="redux" />
          <button
            type="submit">
            Search
          </button>
        </form>
      </Search>
    </div>
    <Table
      list={Array []}
      onDismiss={[Function]}
      pattern="redux">
      <div
        className="table">
        <TableHeader
          activeSortKey="NONE"
          isSortReverse={false}
          onSort={[Function]}>
          <div
            className="table-header">
            <span
              className="midColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="TITLE">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Title
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="midColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="AUTHOR">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Author
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="smallColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="COMMENTS">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Comments
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="smallColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="POINTS">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Points
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="smallColumn">
              <Button
                className="">
                <button
                  className=""
                  type="button">
                  Archive
                </button>
              </Button>
            </span>
          </div>
        </TableHeader>
        <div
          className="table-body" />
      </div>
    </Table>
    <div
      className="interactions">
      <Component
        isLoading={true}
        onClick={[Function]}>
        <Loading>
          <div>
            Loading...
          </div>
        </Loading>
      </Component>
    </div>
  </div>
</App>
`;

exports[`App snapshots 2`] = `
<App
  fetchInitialData={[Function]}
  isLoading={false}>
  <div
    className="page">
    <div
      className="interactions">
      <Search
        onChange={[Function]}
        onSubmit={[Function]}
        value="redux">
        <form
          onSubmit={[Function]}>
          <input
            onChange={[Function]}
            type="text"
            value="redux" />
          <button
            type="submit">
            Search
          </button>
        </form>
      </Search>
    </div>
    <Table
      list={Array []}
      onDismiss={[Function]}
      pattern="redux">
      <div
        className="table">
        <TableHeader
          activeSortKey="NONE"
          isSortReverse={false}
          onSort={[Function]}>
          <div
            className="table-header">
            <span
              className="midColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="TITLE">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Title
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="midColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="AUTHOR">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Author
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="smallColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="COMMENTS">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Comments
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="smallColumn">
              <Sort
                activeSortKey="NONE"
                isSortReverse={false}
                onSort={[Function]}
                sortKey="POINTS">
                <Button
                  className="button-inline"
                  onClick={[Function]}>
                  <button
                    className="button-inline"
                    onClick={[Function]}
                    type="button">
                    Points
                    <span
                      className="fa" />
                  </button>
                </Button>
              </Sort>
            </span>
            <span
              className="smallColumn">
              <Button
                className="">
                <button
                  className=""
                  type="button">
                  Archive
                </button>
              </Button>
            </span>
          </div>
        </TableHeader>
        <div
          className="table-body" />
      </div>
    </Table>
    <div
      className="interactions">
      <Component
        isLoading={true}
        onClick={[Function]}>
        <Loading>
          <div>
            Loading...
          </div>
        </Loading>
      </Component>
    </div>
  </div>
</App>
`;
